post_processing_stages/object_classify_tf_stage.cpp

```diff
@@ -0,0 +1,370 @@
+
+class ObjectClassifyTfStage : public PostProcessingStage
+{
+public:
+	ObjectClassifyTfStage(LibcameraApp *app) : PostProcessingStage(app) {}
+
+	char const *Name() const override;
+
+	void Read(boost::property_tree::ptree const &params);
+
+	void Configure();
+
+	bool Process(CompletedRequest &completed_request);
+
+	void Stop();
+
+private:
+	void initialize();
+
+	void runInference();
+
+	TfLiteStatus readLabelsFile(const tflite::string &file_name, std::vector<tflite::string> *result,
+								size_t *found_label_count);
+
+	std::vector<uint8_t> yuvToRgb(uint8_t *mem, int w, int h, int stride);
+
+	void getTopResults(uint8_t *prediction, int prediction_size, size_t num_results,
+					   std::vector<std::pair<float, int>> *top_results, TfLiteType input_type);
+
+	Stream *stream_;
+	int w_, h_, stride_;
+	std::mutex future_ptr_mutex_;
+	std::mutex output_results_mutex_;
+	TfSettings settings_;
+	std::unique_ptr<tflite::FlatBufferModel> model_;
+	std::unique_ptr<tflite::Interpreter> interpreter_;
+	const std::vector<int> inputs_;
+	const std::vector<int> outputs_;
+	int input_;
+	std::vector<std::pair<std::string, float>> output_results_;
+	std::unique_ptr<std::future<void>> future_ptr_;
+	int refresh_rate_;
+	int top_n_results_;
+	float threshold_high_;
+	float threshold_low_;
+	std::string model_file_;
+	std::string labels_file_;
+	int display_labels_;
+	std::vector<tflite::string> labels_;
+	size_t label_count_;
+	std::vector<uint8_t> tensor_input_;
+	std::vector<std::pair<float, int>> top_results_;
+};
+
+#define NAME "object_classify_tf"
+
+char const *ObjectClassifyTfStage::Name() const
+{
+	return NAME;
+}
+
+void ObjectClassifyTfStage::Read(boost::property_tree::ptree const &params)
+{
+	top_n_results_ = params.get<int>("top_n_results", 3);
+	refresh_rate_ = params.get<int>("refresh_rate", 10);
+	threshold_high_ = params.get<float>("threshold_high", 0.01f);
+	threshold_low_ = params.get<float>("threshold_low", 0.01f);
+	model_file_ = params.get<std::string>("model_file", "/home/pi/models/mobilenet_quant_v1_224.tflite");
+	labels_file_ = params.get<std::string>("labels_file", "/home/pi/models/labels.txt");
+	display_labels_ = params.get<int>("display_labels", 1);
+
+	settings_.number_of_results = top_n_results_;
+	settings_.model_name = model_file_;
+	settings_.labels_file_name = labels_file_;
+
+	initialize();
+}
+
+void ObjectClassifyTfStage::initialize()
+{
+	model_ = tflite::FlatBufferModel::BuildFromFile(settings_.model_name.c_str());
+
+	settings_.model = model_.get();
+	std::cout << "Loaded model " << settings_.model_name << std::endl;
+	model_->error_reporter();
+
+	tflite::ops::builtin::BuiltinOpResolver resolver;
+
+	tflite::InterpreterBuilder(*model_, resolver)(&interpreter_);
+	if (!interpreter_)
+		throw std::runtime_error("Failed to construct interpreter");
+
+	std::cout << "Input size " << interpreter_->inputs().size() << std::endl;
+	interpreter_->SetAllowFp16PrecisionForFp32(settings_.allow_fp16);
```

Comment:
Perhaps we have a "log" option in the settings and only do the std::cout if that is set?
I don't think we want to write to stdout unconditionally.

Reply:
I don't mind too much here as it's a one-off, but I agree with your later similar one.